core/mintController.js

```diff
@@ -1,50 +1,57 @@
 import { makePrivateName } from '../util/PrivateName';
 
 export function makeBasicMintController() {
-  // Map from purse or payment to the rights it currently
-  // holds. Rights can move via payments
-
-  // purse/payment to amount
-  let rights = makePrivateName();
-
-  function recordPayment(src, payment, amount, srcNewRightsAmount) {
-    rights.set(src, srcNewRightsAmount);
-    rights.init(payment, amount);
+  // An asset can either be a purse or payment. An asset controller
+  // keeps track of either all of the purses (purseController) or all
+  // of the payments (paymentController).
+  function makeAssetController(type) {
+    // asset to amount
+    let assets = makePrivateName();
+    return {
+      updateAmount(asset, newAmount) {
+        assets.set(asset, newAmount);
+      },
+      recordNew(asset, initialAmount) {
+        assets.init(asset, initialAmount);
+      },
+      getAmount(asset) {
+        return assets.get(asset);
+      },
+      getType() {
+        return `${type}`;
+      },
+      has(asset) {
+        return assets.has(asset);
+      },
+      destroyAll() {
+        assets = makePrivateName(); // reset completely
+      },
+    };
   }
 
+  // destroy is outside of an assetController because it could affect
+  // purses or payments
   function destroy(_amount) {
     throw new Error('destroy is not implemented');
   }
 
-  function destroyAll() {
-    rights = makePrivateName(); // reset rights
-  }
-
-  function recordDeposit(
-    srcPayment,
-    srcNewRightsAmount,
-    purse,
-    purseNewRightsAmount,
-  ) {
-    rights.set(srcPayment, srcNewRightsAmount);
-    rights.set(purse, purseNewRightsAmount);
-  }
+  const purseController = makeAssetController('purse');
+  const paymentController = makeAssetController('payment');
```

Comment:
I really like that there's no code difference between a purseController and a paymentController.

Reply:
Awesome!